### DISCUSSION TITLE
Use gitversion when promoting builds across environments

### DISCUSSION THREAD

**Discussion by sam-wheat** [GitTools/GitVersion#3428] [Q&A] [answered]
We develop an internal desktop app for a company.  We (roughly) use GitHubFlow.

Developer creates a feature branch, merges back to main. 

When several features are ready to be released on a QA build, developer creates a tag based on main: i.e. `1.2.3-alpha`.  

When a QA build is ready to be promoted to UAT a UAT tag is created based on the QA tag: `1.2.3-beta` based on `1.2.3-alpha`.

Promoting a UAT build to prod is same as preceding.  Manually creating a tag for either QA or UAT is what kicks off the build.

Can we use gitversion for this work flow?  We can change whatever needs to change to maximize the utility of gitversion.  With that said the concept of promoting a build is integral to our workflow.

Only must-have is that we stamp assemblies with a tag that allows us to find the tag it was built from (-alpha = QA, -beta = UAT).

Feature branches are the only branches we create.  Everything else is tags.  I noticed 
the [configuration docs](https://gitversion.net/docs/reference/configuration) discusses Branch Configuration and documents sections for `branches:`.  
Is there similar functionality for tags?  In a perfect world we would name our tags `alpha/1.2.3` and `beta/1.2.3` just to keep them grouped.  Can gitversion use a regex to get the version?

**Comment by HHobeck** [comment, marked as answer]
If I understand you correct you would like to have a different pre-release label dependent on the tag:

```csharp
   [Test]
    public void Just_A_Test()
    {
        var configuration = GitHubFlowConfigurationBuilder.New
            .WithBranch("main", branchBuilder => branchBuilder
                .WithIncrement(IncrementStrategy.Minor)
            ).Build();

        using var fixture = new EmptyRepositoryFixture("main");

        fixture.Repository.MakeACommit();

        fixture.AssertFullSemver("0.1.0+1", configuration);

        fixture.Repository.ApplyTag("0.1.0-alpha.1");

        fixture.AssertFullSemver("0.1.0+1", configuration);

        fixture.Repository.MakeACommit();

        fixture.AssertFullSemver("0.1.0+2", configuration);

        fixture.Repository.ApplyTag("0.1.0-alpha.2");

        fixture.AssertFullSemver("0.1.0+2", configuration);

        fixture.Repository.ApplyTag("0.1.0-beta.1");

        fixture.AssertFullSemver("0.1.0+2", configuration);

        fixture.Repository.ApplyTag("0.1.0");

        fixture.AssertFullSemver("0.1.0", configuration);

        fixture.Repository.MakeACommit();

        fixture.AssertFullSemver("0.2.0+1", configuration);
    }
```

**Reply by sam-wheat** [reply]
Hello Hardy, thank you for your response.


    fixture.Repository.ApplyTag("0.1.0-beta.1");  // base this one on 0.1.0-alpha.2

I want to be able to do the following

create a branch off main
check in some changes
create a tag  // gitversion should allow me to create something like 1.2.3-alpha

The build system will kick off a build based on the new tag.  The new build will go to QA.  

When QA approves the build I want to be able to tell gitversion to promote it to UAT via a new tag.  So I want to tell gitversion something like this: "Create a tag based on `1.2.3-alpha`.  gitversion should know to increment the pre and create a new tag named `1.2.3-beta`.  
Same process for `-beta` which will eventually become `1.2.3`.

**Reply by HHobeck** [reply]
I think this is a better example:

```csharp
    [Test]
    public void Just_Another_Test()
    {
        var configuration = GitHubFlowConfigurationBuilder.New
            .WithBranch("main", branchBuilder => branchBuilder
                .WithIncrement(IncrementStrategy.Minor).WithLabel("develop")
            ).Build();

        using var fixture = new EmptyRepositoryFixture("main");

        fixture.Repository.MakeATaggedCommit("0.2.0-alpha.1");

        fixture.AssertFullSemver("0.2.0-alpha.1", configuration);

        fixture.Repository.MakeACommit();

        fixture.AssertFullSemver("0.2.0-develop.1+1", configuration);

        fixture.Repository.ApplyTag("0.2.0-alpha.2");

        fixture.AssertFullSemver("0.2.0-alpha.2", configuration);

        fixture.Repository.ApplyTag("0.2.0-beta.1");

        fixture.AssertFullSemver("0.2.0-beta.1", configuration);

        fixture.Repository.ApplyTag("0.2.0");

        fixture.AssertFullSemver("0.2.0", configuration);

        fixture.Repository.MakeACommit();

        fixture.AssertFullSemver("0.3.0-develop.1+1", configuration);
    }
```

**Reply by HHobeck** [reply]
> Hello Hardy, thank you for your response.
> 
> ```
> fixture.Repository.ApplyTag("0.1.0-beta.1");  // base this one on 0.1.0-alpha.2
> ```
> 
> I want to be able to do the following
> 
> create a branch off main check in some changes create a tag // gitversion should allow me to create something like 1.2.3-alpha
> 
> The build system will kick off a build based on the new tag. The new build will go to QA.
> 
> When QA approves the build I want to be able to tell gitversion to promote it to UAT via a new tag. So I want to tell gitversion something like this: "Create a tag based on `1.2.3-alpha`. gitversion should know to increment the pre and create a new tag named `1.2.3-beta`. Same process for `-beta` which will eventually become `1.2.3`.

I'm not certain if I understood your workflow correct. For me it makes no sense to tag a version, build an artifact and deploy the software to any environment. You should ask yourself for what reason GitVersion is good for: It's there to automatically detect the semantic version based on the git repository state. Thus better building the artifact first, deploy to an environment and tag the version afterwards.

If you have a alpha and beta release stage then the git flow is maybe the right workflow for you. Please take a look to the following sequence diagram: https://gitversion.net/docs/learn/branching-strategies/gitflow/examples#minor-release-branches

**Reply by sam-wheat** [reply]
Thank you Hardy.
